pilot/adapter/serviceregistry/aggregate/controller.go

```diff
@@ -15,167 +15,527 @@
+	labelServiceName = labelServicePrefix + "name"
+	// xDS external interface for dns name
+	labelServiceExternalName = labelServicePrefix + "externalName"
+	// xDS external interface for service VIP
+	labelServiceVIP     = labelServicePrefix + "vip"
+	labelInstancePrefix = labelXdsPrefix + "ServiceInstance."
+	// xDS external interface for the instance IP.
+	// Exernal format is a valid ipv4/v6 string format, ex: 10.1.1.3
+	// Internally stored format is a hex representation
+	labelInstanceIP = labelInstancePrefix + "ip"
+	// xDS external interface for instance Port
+	// Exernal format is a valid 16 bit unsigned integer
+	// Internally stored format is a hex representation
+	labelInstancePort      = labelInstancePrefix + "port"
+	labelInstanceNamedPort = labelInstancePrefix + "namedPort"
+
+	// Value is Service Name + ip + port
+	labelInstanceRef = labelInstancePrefix + "reference"
+)
+
 // Registry specifies the collection of service registry related interfaces
 type Registry struct {
 	Name platform.ServiceRegistry
 	model.Controller
-	model.ServiceDiscovery
-	model.ServiceAccounts
+	// The service mesh view that this registry belongs to
+	MeshView *MeshResourceView
 }
 
-// Controller aggregates data across different registries and monitors for changes
-type Controller struct {
+// MeshResourceView is an aggregated store for resources sourced from various registries
+type MeshResourceView struct {
 	registries []Registry
+
+	// Mutex guards services, serviceInstances and corresponding labels
+	mu sync.RWMutex
+
+	// Canonical map of mesh service key to service references
+	services map[resourceKey]*model.Service
+
+	// Canonical map of mesh service instance keys to service instance references
+	serviceInstances map[resourceKey]*model.ServiceInstance
+
+	// A reverse map that associates label names to label values and their associated service resource keys
+	serviceLabels nameValueKeysMap
+
+	// A reverse map that associates label names to label values and their associated service instance resource keys
+	serviceInstanceLabels nameValueKeysMap
+
+	// Cache eviction handler for notifying changes to model objects
+	cacheEvictionHandler model.CacheEvictionHandler
 }
 
-// NewController creates a new Aggregate controller
-func NewController() *Controller {
-	return &Controller{
-		registries: make([]Registry, 0),
+type serviceList []*model.Service
+
+func (sl *serviceList) String() string {
+	out := "["
+	for _, s := range *sl {
+		out = out + s.Hostname + ", "
 	}
+	out = out + "]"
+	return out
 }
 
-// AddRegistry adds registries into the aggregated controller
-func (c *Controller) AddRegistry(registry Registry) {
-	c.registries = append(c.registries, registry)
+type serviceInstanceList []*model.ServiceInstance
+
+func (sl *serviceInstanceList) String() string {
+	out := "["
+	for _, i := range *sl {
+		out = out + i.Service.Hostname + ":" + i.Endpoint.Address + ":" + fmt.Sprintf("%d", i.Endpoint.Port) + ", "
```

Comment:
This is going to be  avery very long string, if I have 100K instances.. What do we need this for?